### PR TITLE
Some fixes and cleanup

### PR DESCRIPTION
**Pull request by DaRacci**

### CODE REVIEW

**Comment by DaRacci** [comment]
Also I'm looking at hopefully getting to the Minimessage support some time this week

File: AdvancedTeleport-Bukkit/src/main/java/io/github/niestrat99/advancedteleport/api/ATPlayer.java
Comment (Thatsmusic99) [inline AdvancedTeleport-Bukkit/src/main/java/io/github/niestrat99/advancedteleport/api/ATPlayer.java:182]: I changed this from Boolean to Void to have exceptions properly handled within the handleAsync function, the Boolean just indicates whether the operation was successful (true if no errors, false if no errors). Would you recommend it any particular way?

Comment (Thatsmusic99) [reply AdvancedTeleport-Bukkit/src/main/java/io/github/niestrat99/advancedteleport/api/ATPlayer.java:182]: Same goes for any other method signatures changed to reflect this.

File: AdvancedTeleport-Bukkit/src/main/java/io/github/niestrat99/advancedteleport/config/NewConfig.java
Comment (Thatsmusic99) [inline AdvancedTeleport-Bukkit/src/main/java/io/github/niestrat99/advancedteleport/config/NewConfig.java:69]: The ConfigSection interface should be used here instead of the implementation (CMConfigSection), is there a reason this cannot be used?

File: AdvancedTeleport-Bukkit/src/main/java/io/github/niestrat99/advancedteleport/payments/PaymentManager.java
Comment (Thatsmusic99) [inline AdvancedTeleport-Bukkit/src/main/java/io/github/niestrat99/advancedteleport/payments/PaymentManager.java:109]: Still aiming for Java 8 support here for the time being, which does not support var.

**Review by Thatsmusic99 — CHANGES_REQUESTED** [review]
Couldn't comment on gradlew but is that a file that needs ignoring whilst we're taking a look through .gitignore?

Just wanted to add some extra points for input as well including some of the return values in the API and other bits and pieces.

Looking forward to MiniMessage support as well, I'm no doubt a MM simp now with how much I've worked with it in comparison to the legacy codes!

File: AdvancedTeleport-Bukkit/src/main/java/io/github/niestrat99/advancedteleport/config/NewConfig.java
Comment (DaRacci) [reply AdvancedTeleport-Bukkit/src/main/java/io/github/niestrat99/advancedteleport/config/NewConfig.java:69]: Wasn't aware of the interface type, i was just going off the exception and what it was expecting ill have that changed

**Comment by DaRacci** [comment]
Also I'm thinking due to the large scale that 5.7 will be it may be worth outright removing some of the deprecated functions in the api and so on

**Comment by DaRacci** [comment]
Would you be open to using the cloud command framework to replace the current commands as its much cleaner and easier to work with

File: AdvancedTeleport-Bukkit/src/main/java/io/github/niestrat99/advancedteleport/api/ATPlayer.java
Comment (DaRacci) [reply AdvancedTeleport-Bukkit/src/main/java/io/github/niestrat99/advancedteleport/api/ATPlayer.java]: I've made a kinda simplified way of handling the errors which i think is cleaner and can also allow for proving more detailed error details via the created exceptions.
Honestly i don't have much experience with futures and i mainly work in kotlin with coroutines and the arrow functional framework for representing values that can fail with the either interface.

**Comment by Thatsmusic99** [comment]
Are you happy to send over the repo for the framework/show an example of how it'd look? I'd be more than happy to give my thoughts on it.

As for deprecated methods, go ahead with it and I'll see what's changed. :)

File: AdvancedTeleport-Bukkit/src/main/java/io/github/niestrat99/advancedteleport/payments/PaymentManager.java
Comment (Thatsmusic99) [reply AdvancedTeleport-Bukkit/src/main/java/io/github/niestrat99/advancedteleport/payments/PaymentManager.java:109]: 1.16 is still Java 8, no? I know Paper released their statement on Java 11 around this time but that doesn't include Spigot which is the main platform AT still has to support (for now).

File: AdvancedTeleport-Bukkit/src/main/java/io/github/niestrat99/advancedteleport/api/ATPlayer.java
Comment (Thatsmusic99) [reply AdvancedTeleport-Bukkit/src/main/java/io/github/niestrat99/advancedteleport/api/ATPlayer.java]: At a glance, it looks a lot tidier! Thank you for covering that. 

It's absolutely fine, I believe I started using the method in one of my standalone programs which otherwise resulted in missing exceptions that would lead to botched logic conclusions. It may be slightly different with AT due to the use of the async executor but the more you know.

**Comment by DaRacci** [comment]
> 

https://github.com/Incendo/cloud
for the most clean example this is the framework in the [carbon plugin](https://github.com/Hexaoxide/Carbon/tree/2.1/common/src/main/java/net/draycia/carbon/common/command/commands)
for a one file example there is [this one](https://github.com/DaRacci/Elixir/blob/master/Elixir-Core/src/main/kotlin/dev/racci/elixir/core/services/CommandService.kt) however that is in kotlin and kinda a mess rn

**Comment by DaRacci** [comment]
Theres also some other examples and guides in the plugins repo

**Comment by Thatsmusic99** [comment]
yo that looks badass, I like it

Since it's going to be a lot of work to implement and test, do you want to put it in a separate branch/PR?

**Comment by DaRacci** [comment]
Yeah ill shove it into a new branch once im fixed up on this one, I was thinking of implementing the library requirements using a runtime downloader such as the one i [maintain mainly for my own plugins](https://github.com/DaRacci/slimjar),
It would allow the downloading and (shading for some libraries that some specific developers love to somehow break all plugins which also use them) of plugins at runtime also meaning checking for things like being a on paper server and not downloading adventure

**Comment by Thatsmusic99** [comment]
Yeah, I was poking around with Adventure not too long ago for HeadsPlus and found the implemented libraries downloader completely botched anything and everything. Having a tool like the one you mentioned would definitely be handy.

**Comment by Thatsmusic99** [comment]
On that note I'll be starting work shortly so I'll get back to you on anything else hopefully at the end of the day if I don't need to scream over anything else. :)

**Comment by DaRacci** [comment]
Also what was your decision for the java version with what i mentioned about it, I would very much like to be able to use newer java features and the min version for 5.7 of 1.16 should allow that

File: AdvancedTeleport-Bukkit/src/main/java/io/github/niestrat99/advancedteleport/payments/PaymentManager.java
Comment (DaRacci) [reply AdvancedTeleport-Bukkit/src/main/java/io/github/niestrat99/advancedteleport/payments/PaymentManager.java:109]: I believe 1.16 was fully compatible with java 17, ill have to do some digging to find out

Comment (DaRacci) [reply AdvancedTeleport-Bukkit/src/main/java/io/github/niestrat99/advancedteleport/payments/PaymentManager.java:109]: Okay so 1.16 does indeed support a maximum java version of 16 meaning we could move the java version to 16 features and maintain it on 17

**Comment by DaRacci** [comment]
While going through and doing some more cleanup it seems that the final keyword has already started to be thrown into things like events

File: AdvancedTeleport-Bukkit/src/main/java/io/github/niestrat99/advancedteleport/payments/PaymentManager.java
Comment (Thatsmusic99) [reply AdvancedTeleport-Bukkit/src/main/java/io/github/niestrat99/advancedteleport/payments/PaymentManager.java:109]: Sorry, I'm a little behind since I was with family and working on an event last night. I got off today and will focus more on my personal projects at least. :) 

I am a little opposed to enforcing a minimum version of Java 16 on servers since I work in server hosting and the place I work at automatically sets Java versions depending on the server version and type being used, i.e. Java 11 for Paper 1.16.5, Java 8 for Spigot and Java 8 for Forge 1.16.5 (including this because of Arclight, which I don't actively encourage but don't discourage either). Because of that, I don't fully know how other popular hosting providers work with this, but looking at plugins that go down this road - looking at you, PlotSquared - are a bit of a pain in the ass to deal with for both us and server owners who just want the plugin to work outside of the box.

We can certainly make Java 16 a minimum version when 1.16.5 eventually wilts out like 1.12.2 and 1.8, but I think we can hold off that for a bit longer.

Comment (DaRacci) [reply AdvancedTeleport-Bukkit/src/main/java/io/github/niestrat99/advancedteleport/payments/PaymentManager.java:109]: Hmm, that makes sense for the providers which don't allow changing the java version, I hadn't thought about that as I've only ever hosted on local hardware or full dedicated machines, I'll have to go back and double check what i need to change if anything for java 11 but that should at-least allow for some of the newer features.

Comment (Thatsmusic99) [reply AdvancedTeleport-Bukkit/src/main/java/io/github/niestrat99/advancedteleport/payments/PaymentManager.java:109]: Yeah, I fully understand. It's worth having the extra perspective. I try to aim at making the plugin as easily usable as possible, even if it still needs some cleaning up here and there. Thank you for taking that into account!

Var was introduced in Java 10 iirc.

Comment (DaRacci) [reply AdvancedTeleport-Bukkit/src/main/java/io/github/niestrat99/advancedteleport/payments/PaymentManager.java:109]: So interesting development is that without setting the java version it 16 at the lowest the paper api and some other dependencies become unresolvable and require a higher java version, While we could compile it to 16 with just 11 features it still wouldn't run on java 11 and would require a minimum of 16

Comment (Thatsmusic99) [reply AdvancedTeleport-Bukkit/src/main/java/io/github/niestrat99/advancedteleport/payments/PaymentManager.java:109]: I think I did run into that when initially introducing Gradle to the dev environment for AT, I *think* it ran fine on Java 8? Not quite sure. I do love the mess that is MC's Java version management

Comment (DaRacci) [reply AdvancedTeleport-Bukkit/src/main/java/io/github/niestrat99/advancedteleport/payments/PaymentManager.java:109]: Doesn't seem to work with 8 either, and with more libraries overtime this will become more of an issue

Comment (Thatsmusic99) [reply AdvancedTeleport-Bukkit/src/main/java/io/github/niestrat99/advancedteleport/payments/PaymentManager.java:109]: Crud :( okay, keep Java 16/17 in place then, I'll keep it maintained at the latest server version support (i.e. 1.18 and 1.19). I can maintain a deprecated branch of AT solely for bug fixes in older versions. Sorry for the trouble.

Comment (DaRacci) [reply AdvancedTeleport-Bukkit/src/main/java/io/github/niestrat99/advancedteleport/payments/PaymentManager.java:109]: That's fine, there not really much of a way to work around minecrafts absolutely fucked java versions.

**Comment by DaRacci** [comment]
Marking as ready for review as my brain is broken and it feels like enough for now

**Comment by Thatsmusic99** [comment]
jesus christ

**Comment by DaRacci** [comment]
Marking back to draft had some more i want to do, also with the scale of this update and most likely a ton of API breaking would it be worth making the version 6.0.0 so show how large of an update this is?

**Comment by Thatsmusic99** [comment]
Yeah, I was really thinking about that actually, since it's several *large* updates in one including the API, and by [Semvar's](https://semver.org/) conventions, yeah, incompatible API changes need a major version increase.

**Comment by Thatsmusic99** [comment]
Ohh, that looks like it'd be neat to follow. With Cog I can probably look into working with [Minotaur](https://github.com/modrinth/minotaur) as well since AT is also on Modrinth now. Still a pain to maintain and update on SpigotMC, but aye, that's SpigotMC. 

If you're happy I can try and look into handling this in my own branch so I'm familiar with what it does and how it works.

I do try to follow conventions similar to what Cog specifies at work so chances are I can enforce it with this project as well as sister projects like HP. Ignoring the fact that I tend to go [off-rail with some commit names](https://github.com/Niestrat99/AT-Rewritten/commit/1d6124f6673a5021153a479d38a60794c4102c55).

**Comment by DaRacci** [comment]
sounds good ill leave it to you then, its fairy easy to setup and id also recommend installing [koji](https://github.com/its-danny/koji) which makes creating commits following the format much easier

**Comment by Thatsmusic99** [comment]
Just wanted to follow up with this, since I did want to get v6 out sometime before my next semester starts and I get absorbed by another work project - I'm working on incorporating some of these changes in a separate branch, this being [chore/code-comments](https://github.com/Niestrat99/AT-Rewritten/tree/chore/code-comments). Mainly with respect to annotations but for now I'm keeping the Contracts out of it since I felt like NotNull and Nullable are more suitable at this time. May need a rebase though and I can cover that myself should it be required.

I'll still keep this PR here since it's still got a lot going on but don't burn yourself out with major changes. If you do think anything major needs changing or switching up so PRs don't end up ridiculously massive (I'm fully aware I'm guilty of this myself), feel free to reach out to me directly and I can split the workload where necessary.

**Comment by DaRacci** [comment]
I'm not going to be able to work on this for some time, It may be upwards of a month before i can work on anything, If you want to call this PR done and start merging it thats fine as i can always come back later to cleanup the rest of what i wanted to do.

**Comment by Thatsmusic99** [comment]
Alrighty, I'll go ahead with merging but put it into the branch I mentioned since that's currently my own "cleanup" branch. Is there anything extra you wanted to add with it? I'll be happy to cover that myself in my own PR.

**Comment by DaRacci** [comment]
No there wasn't anything else currently, Just the command framework but thats a job for another time since its a quite large change, and the minimessage is all done too just was waiting on this branch.

**Comment by Thatsmusic99** [comment]
That sounds fine. I'll consider looking into Cloud in my own time in the event that I want to start porting AT to different platforms. MiniMessage I will also review when I get everything merged.

**Comment by Thatsmusic99** [comment]
Hoooo crap that was a pain to rebase and merge.

Anyhoo, on that note, I am running into an issue with the updated Gradle file; I'm fairly certain this is something I have to tweak myself, but libMinix does not want to define itself seemingly whatsoever:

image

I did checkout to the [last commit made directly by you in this branch](https://github.com/DaRacci/AT-Rewritten/tree/5f32faab59ba59bb6d529ec7baeaf6df2af9824a) but with no luck. The extra code for creating the version catalogue under libMinix seems to be in the settings.gradle.kts is intact and untouched, so I'm not sure if I'm missing something crucial. Is there anything extra I need to run to get it set up?

Cheers in advance! Things do look a lot cleaner.

**Comment by DaRacci** [comment]
That's odd i just checked out the latest commit and it resolved for me fine, maybe this will work for you with an updated build?

**Comment by Thatsmusic99** [comment]
Ah yeah, I'm a Windows 11 user, that may as well explain it. The project uses Gradle v7.6 which looks like the latest version too. Just gave it a kick through WSL Ubuntu and it worked fine, but failed when using IntelliJ on Windows and PowerShell. Bloody brilliant.

I'll see what can be done about that.

**Comment by Thatsmusic99** [comment]
Excuse my french, it's a piece of shit, it worked after I changed it to `libsMinix` and won't take `libMinix`, why!?

How much damage would changing the name of the version catalogue do internally?

**Comment by DaRacci** [comment]
Oh and I may be wrong but I think that error is just for the intellij stuff, you should be able to compile it with gradle fine

**Comment by Thatsmusic99** [comment]
Caches would make sense - giving those a wipe now, sorry about the outburst lmao, WSL was already antagonising me when I tried to get it set up

With regards to it being IntelliJ only, it doesn't seem to be unfortunately, I tried building and refreshing dependencies through the PS terminal but it wouldn't define the dependencies there either. I've also got it set up to use Gradle to run and build tasks rather than get IntelliJ involved, so even if it was just IntelliJ, running Gradle through it would've yielded differing results. Once the cache is gone though, I'll see what happens.

**Comment by Thatsmusic99** [comment]
Cache clearing has done the trick, thank you for pointing that out! Glad it was an easy fix and not something bizarre and convoluted.

I'll work on clearing up any remaining TODOs from this PR before merging it.

**Comment by DaRacci** [comment]
> Caches would make sense - giving those a wipe now, sorry about the outburst lmao, WSL was already antagonising me when I tried to get it set up
> 
> With regards to it being IntelliJ only, it doesn't seem to be unfortunately, I tried building and refreshing dependencies through the PS terminal but it wouldn't define the dependencies there either. I've also got it set up to use Gradle to run and build tasks rather than get IntelliJ involved, so even if it was just IntelliJ, running Gradle through it would've yielded differing results. Once the cache is gone though, I'll see what happens.

Yeah I tried with WSL like 2 years back and just decided to make the full move to linux because just was sick of windows shit and wsl wasn't any better.

**Comment by Thatsmusic99** [comment]
Yup, I vanished, my SSD has been throwing a fit and as predicted, I did get absorbed into that work project again. :) I believe the majority of things should be done now so I can merge this and start focusing on introducing MiniMessage.

Before I do that though, is there any input you want on any tweaks made? I've pretty much limited it to consistency and compilation errors but some things I will be following up with on the base branch.

**Comment by DaRacci** [comment]
Nah looks all good to me, merge at will.

**Comment by Thatsmusic99** [comment]
Epic. Thank you again!

**Pull request merged**
